packages/utils/node-resolver-rs/src/lib.rs

```diff
@@ -0,0 +1,2479 @@
+        if self.specifier_type == SpecifierType::Url {
+          Ok(Resolution::External)
+        } else {
+          let (scheme, _) = parse_scheme(url)?;
+          Err(ResolverError::UnknownScheme {
+            scheme: scheme.into_owned(),
+          })
+        }
+      }
+      _ => Err(ResolverError::UnknownError),
+    }
+  }
+
+  fn find_ancestor_file(&self, from: &Path, filename: &str) -> Option<PathBuf> {
+    let from = from.parent().unwrap();
+    self
+      .resolver
+      .find_ancestor_file(from, filename, &self.invalidations)
+  }
+
+  fn find_package(&self, from: &Path) -> Result<Option<&'a PackageJson<'a>>, ResolverError> {
+    self.resolver.find_package(from, &self.invalidations)
+  }
+
+  fn resolve_relative(&self, specifier: &Path, from: &Path) -> Result<Resolution, ResolverError> {
+    // Resolve aliases from the nearest package.json.
+    let path = resolve_path(from, specifier);
+    let package = if self.resolver.flags.contains(Flags::ALIASES) {
+      self.find_package(&path.parent().unwrap())?
+    } else {
+      None
+    };
+
+    if let Some(res) = self.load_path(&path, package)? {
+      return Ok(res);
+    }
+
+    Err(ResolverError::FileNotFound {
+      relative: specifier.to_owned(),
+      from: from.to_owned(),
+    })
+  }
+
+  fn resolve_bare(&self, module: &str, subpath: &str) -> Result<Resolution, ResolverError> {
+    let include = match self.resolver.include_node_modules.as_ref() {
+      IncludeNodeModules::Bool(b) => *b,
+      IncludeNodeModules::Array(a) => a.iter().any(|v| v == module),
+      IncludeNodeModules::Map(m) => *m.get(module).unwrap_or(&true),
+    };
+
+    if !include {
+      return Ok(Resolution::External);
+    }
+
+    if self.resolver.flags.contains(Flags::ALIASES) {
+      // First, check for an alias in the root package.json.
+      if let Some(package) = self.root_package()? {
+        if let Some(res) = self.resolve_aliases(
+          package,
+          &Specifier::Package(Cow::Borrowed(module), Cow::Borrowed(subpath)),
+          Fields::ALIAS,
+        )? {
+          return Ok(res);
+        }
+      }
+
+      // Next, try the local package.json.
+      if let Some(package) = self.find_package(&self.from.parent().unwrap())? {
+        let mut fields = Fields::ALIAS;
+        if self.resolver.entries.contains(Fields::BROWSER) {
+          fields |= Fields::BROWSER;
+        }
+        if let Some(res) = self.resolve_aliases(
+          package,
+          &Specifier::Package(Cow::Borrowed(module), Cow::Borrowed(subpath)),
+          fields,
+        )? {
+          return Ok(res);
+        }
+      }
+    }
+
+    // Next, check tsconfig.json for the paths and baseUrl options.
+    if let Some(res) = self.resolve_tsconfig_paths()? {
+      return Ok(res);
+    }
+
+    self.resolve_node_module(module, subpath)
+  }
+
+  fn resolve_node_module(&self, module: &str, subpath: &str) -> Result<Resolution, ResolverError> {
+    // TODO: check if module == self
```

Comment:
Still relevant?

Reply:
The [node docs](https://nodejs.org/api/packages.html#self-referencing-a-package-using-its-name) have a section on this, and it's even got a special [PACKAGE_SELF_RESOLVE](https://nodejs.org/api/esm.html#resolver-algorithm-specification) function in the spec. But I don't really understand why it's called out specifically, because self references work even without this special handling just by searching the node_modules folders as with any other package. I added some tests for it but 🤷